src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/AppointmentEventPublisher.kt

```diff
@@ -24,26 +33,55 @@ class AppointmentEventPublisher(
   private val applicationEventPublisher: ApplicationEventPublisher,
   private val locationMapper: LocationMapper
 ) {
-  fun attendanceRecordedEvent(session: ActionPlanSession, notifyPP: Boolean) {
+  fun attendanceRecordedEvent(appointment: Appointment, notifyPP: Boolean, appointmentType: AppointmentType) {
     applicationEventPublisher.publishEvent(
-      AppointmentEvent(this, AppointmentEventType.ATTENDANCE_RECORDED, session, getAppointmentURL(session), notifyPP)
+      AppointmentEvent(
+        this,
+        AppointmentEventType.ATTENDANCE_RECORDED,
+        appointment,
+        getAppointmentURL(appointment, appointmentType),
+        notifyPP,
+        appointmentType
+      )
     )
   }
 
-  fun behaviourRecordedEvent(session: ActionPlanSession, notifyPP: Boolean) {
+  fun behaviourRecordedEvent(appointment: Appointment, notifyPP: Boolean, appointmentType: AppointmentType) {
     applicationEventPublisher.publishEvent(
-      AppointmentEvent(this, AppointmentEventType.BEHAVIOUR_RECORDED, session, getAppointmentURL(session), notifyPP)
+      AppointmentEvent(
+        this,
+        AppointmentEventType.BEHAVIOUR_RECORDED,
+        appointment,
+        getAppointmentURL(appointment, appointmentType),
+        notifyPP,
+        appointmentType
+      )
     )
   }
 
-  fun sessionFeedbackRecordedEvent(session: ActionPlanSession, notifyPP: Boolean) {
+  fun sessionFeedbackRecordedEvent(appointment: Appointment, notifyPP: Boolean, appointmentType: AppointmentType) {
     applicationEventPublisher.publishEvent(
-      AppointmentEvent(this, AppointmentEventType.SESSION_FEEDBACK_RECORDED, session, getAppointmentURL(session), notifyPP)
+      AppointmentEvent(
+        this,
+        AppointmentEventType.SESSION_FEEDBACK_RECORDED,
+        appointment,
+        getAppointmentURL(appointment, appointmentType),
+        notifyPP,
+        appointmentType
+      )
     )
   }
 
-  private fun getAppointmentURL(session: ActionPlanSession): String {
-    val path = locationMapper.getPathFromControllerMethod(ActionPlanSessionController::getSession)
-    return locationMapper.expandPathToCurrentRequestBaseUrl(path, session.actionPlan.id, session.sessionNumber).toString()
+  private fun getAppointmentURL(appointment: Appointment, appointmentType: AppointmentType): String {
+    when (appointmentType) {
+      AppointmentType.SERVICE_DELIVERY -> run {
+        NotifyAppointmentService.logger.error("action plan session should not be using the shared appointment notify service.")
+        return ""
```

Comment:
Does an empty string cause a problem later on?

Reply:
I'm not sure this is even used? Service_delivery appointments should never reach this code until the refactor is done. If they did for whatever reason, they should be stopped inside the notify service

Reply:
Domain events need to provide a callback URL where subscribers can interrogate the details of what happened -- this is used for that, but it's getting lost in translation. Some refactoring is due to make these responsibilities clearer